sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java

```diff
@@ -440,6 +442,75 @@ public BoundedWindow window() {
   }
 
   /**
+    * Annotation for declaring and dereferencing timers.
+    *
+    * <p><i>Not currently supported by any runner.</i>
+    *
+    * <p>To declare a timer, create a field of type {@link TimerSpec} annotated with a {@link
+    * TimerId}. To use the cell during processing, add a parameter of the appropriate {@link Timer}
+    * subclass to your {@link ProcessElement @ProcessElement} method, and annotate it with {@link
+    * TimerId}. See the following code for an example:
+    *
+    * <pre>{@code
+    * new DoFn<KV<Key, Foo>, Baz>() {
+    *   @TimerId("my-timer-id")
+    *   private final TimerSpec myTimer = TimerSpecs.timerForDomain(TimeDomain.EVENT_TIME);
+    *
+    *   @ProcessElement
+    *   public void processElement(
+    *       ProcessContext c,
+    *       @TimerId("my-timer-id") Timer myTimer) {
```

Comment:
Just a thought: do we actually need the field? We already repeat the timer name `"my-timer-id"` twice - in the `ProcessElement` parameter and in the `onMyTimer` callback, why have it a third time? And we have both the field and the parameter.

Can we have something like:
`void processElement(..., @Timer("my-timer-id", EVENT_TIME) Timer myTimer)` or even `myTimer.setForNowPlus(duration, EVENT_TIME)`?